File: ui/src/common/types/miniAppPlatform.ts
/**
 * @license
 * Copyright 2025-2026 NomiFun (nomifun.com)
 * SPDX-License-Identifier: Apache-2.0
 */

import type { MiniAppId } from './ids';
import type { CapabilityCatalogItem } from './agentPlatform/contracts';

export type { MiniAppId } from './ids';

export type MiniAppKind = 'ui_only' | 'service';
export type MiniAppServiceLifecycle = 'on_demand' | 'continuous';
export type MiniAppLifecycle = 'enabled' | 'disabled' | 'trashed' | 'deleting';
export type MiniAppPublishMode = 'manual' | 'auto_ui_only';
export type MiniAppProjectSourceState = 'empty' | 'editable' | 'runtime_only';
export type MiniAppTestStatus =
  | 'not_required'
  | 'not_run'
  | 'passed'
  | 'failed'
  | 'needs_test_input'
  | 'stale';

export type MiniAppServiceHealth =
  | { state: 'not_applicable' }
  | { state: 'stopped' }
  | {
      state: 'starting';
      release_id: string;
      expected_release_digest: string;
    }
  | {
      state: 'ready';
      release_id: string;
      expected_release_digest: string;
      started_at_ms: number;
    }
  | {
      state: 'failed';
      release_id: string;
      expected_release_digest: string;
      error_code: string;
    };

export interface MiniAppReleaseRef {
  release_id: string;
  artifact_id: string;
  release_digest: string;
  manifest_digest: string;
}

export interface MiniAppReleasePointers {
  pointer_revision: number;
  active_release_epoch: number;
  ready?: MiniAppReleaseRef;
  active?: MiniAppReleaseRef;
  previous?: MiniAppReleaseRef;
}

export interface MiniAppSummary {
  miniapp_id: MiniAppId;
  product_revision: number;
  display_name: string;
  description?: string;
  icon_asset_id?: string;
  kind: MiniAppKind;
  lifecycle: MiniAppLifecycle;
  releases: MiniAppReleasePointers;
  service_health: MiniAppServiceHealth;
  surface_available: boolean;
  updated_at_ms: number;
}

export interface MiniAppLibraryResponse {
  library_revision: number;
  miniapps: MiniAppSummary[];
}

export interface MiniAppServiceDescriptor {
  lifecycle: MiniAppServiceLifecycle;
  uses_files: boolean;
  uses_private_database: boolean;
  service_contract_digest: string;
}

export interface MiniAppReleaseTest {
  status: MiniAppTestStatus;
  release_id: string;
  expected_release_digest: string;
  receipt_id?: string;
  expected_service_run_key?: string;
  issued_at_ms?: number;
  error_code?: string;
}

export interface MiniAppReadyRelease {
  release: MiniAppReleaseRef;
  project_build_generation: number;
  created_at_ms: number;
  kind: MiniAppKind;
  service?: MiniAppServiceDescriptor;
  test: MiniAppReleaseTest;
  migration_count: number;
  can_publish: boolean;
  can_auto_publish: boolean;
  blocking_reasons: string[];
}

export type MiniAppCapabilityContribution = CapabilityCatalogItem;

export interface MiniAppConfigSchema {
  schema_digest: string;
  schema: unknown;
}

export interface MiniAppConfigState {
  config_revision: number;
  schema_digest: string;
  values: unknown;
  valid: boolean;
  validation_errors: string[];
}

export type MiniAppCredentialBindingStatus =
  | 'unbound'
  | 'bound'
  | 'missing'
  | 'invalid';

export interface MiniAppCredentialSlotBinding {
  slot_key: string;
  display_name: string;
  required: boolean;
  status: MiniAppCredentialBindingStatus;
  credential_id?: string;
}

export type MiniAppOperationKind =
  | 'build'
  | 'import'
  | 'export'
  | 'miniapp_permanent_delete';
export type MiniAppOperationState =
  | 'running'
  | 'succeeded'
  | 'failed'
  | 'canceled';
export type MiniAppOperationOwner =
  | { owner: 'plugin_project'; project_id: string }
  | { owner: 'plugin_mount'; mount_id: string }
  | { owner: 'miniapp'; miniapp_id: MiniAppId };

export interface MiniAppOperationSummary {
  operation_id: string;
  operation_revision: number;
  kind: MiniAppOperationKind;
  owner: MiniAppOperationOwner;
  state: MiniAppOperationState;
  cancelable: boolean;
  progress_percent?: number;
  started_at_ms: number;
  completed_at_ms?: number;
}

export interface MiniAppWorkshop {
  miniapp: MiniAppSummary;
  service_lifecycle?: MiniAppServiceLifecycle;
  active_service?: MiniAppServiceDescriptor;
  project_id: string;
  project_revision: number;
  publish_mode: MiniAppPublishMode;
  source_state: MiniAppProjectSourceState;
  build_generation: number;
  source_snapshot_digest?: string;
  dependency_lock_digest?: string;
  ready?: MiniAppReadyRelease;
  config_schema: MiniAppConfigSchema;
  config: MiniAppConfigState;
  credential_bindings_revision: number;
  credential_slots: MiniAppCredentialSlotBinding[];
  capabilities: MiniAppCapabilityContribution[];
  active_operation?: MiniAppOperationSummary;
}

export interface CreateMiniAppProjectRequest {
  expected_library_revision: number;
  display_name: string;
  description?: string;
  kind: MiniAppKind;
}

export interface MiniAppSourceFile {
  miniapp_id: MiniAppId;
  project_id: string;
  path: string;
  content: string;
  source_snapshot_digest: string;
  build_generation: number;
}

export interface GetMiniAppSourceFileRequest {
  miniapp_id: MiniAppId;
  path: string;
}

export interface ReplaceMiniAppSourceFileRequest {
  miniapp_id: MiniAppId;
  expected_product_revision: number;
  project_id: string;
  expected_project_revision: number;
  expected_build_generation: number;
  expected_source_snapshot_digest: string;
  path: string;
  content: string;
}

export interface BuildMiniAppRequest {
  miniapp_id: MiniAppId;
  expected_product_revision: number;
  project_id: string;
  expected_project_revision: number;
  expected_build_generation: number;
  expected_source_snapshot_digest: string;
  expected_dependency_lock_digest: string;
  service_lifecycle?: MiniAppServiceLifecycle;
}

export interface TestMiniAppReleaseRequest {
  miniapp_id: MiniAppId;
  expected_product_revision: number;
  expected_pointer_revision: number;
  project_id: string;
  expected_project_revision: number;
  expected_build_generation: number;
  release_id: string;
  expected_release_digest: string;
  expected_config_revision: number;
  expected_credential_bindings_revision: number;
  resolved_test_input_digest: string;
}

export type MiniAppShareContent = 'ready_release' | 'active_release';

export interface ShareMiniAppRequest {
  miniapp_id: MiniAppId;
  expected_product_revision: number;
  expected_pointer_revision: number;
  content: MiniAppShareContent;
  release_id: string;
  expected_release_digest: string;
  destination_path: string;
  include_source: boolean;
}

export interface ImportMiniAppShareRequest {
  expected_library_revision: number;
  source_path: string;
  expected_bundle_digest: string;
  expected_release_digest: string;
  display_name: string;
}

export interface ImportMiniAppArtifactRequest {
  expected_library_revision: number;
  source_path: string;
  expected_artifact_digest: string;
  display_name: string;
}

export interface ExportMiniAppBackupRequest {
  miniapp_id: MiniAppId;
  expected_product_revision: number;
  expected_lifecycle: 'disabled';
  expected_pointer_revision: number;
  expected_config_revision: number;
  expected_credential_bindings_revision: number;
  destination_path: string;
}

export interface ImportMiniAppBackupRequest {
  expected_library_revision: number;
  source_path: string;
  expected_backup_metadata_digest: string;
  display_name: string;
}

export interface CancelMiniAppBuildRequest {
  miniapp_id: MiniAppId;
  operation_id: string;
  expected_operation_revision: number;
}

export interface PublishMiniAppRequest {
  miniapp_id: MiniAppId;
  expected_product_revision: number;
  expected_pointer_revision: number;
  expected_active_release_epoch: number;
  ready_release_id: string;
  expected_ready_release_digest: string;
  expected_active_release_digest?: string;
  expected_service_test_receipt_id?: string;
  acknowledge_test_warning: boolean;
}

export interface RollbackMiniAppRequest {
  miniapp_id: MiniAppId;
  expected_product_revision: number;
  expected_pointer_revision: number;
  expected_active_release_epoch: number;
  expected_current_release_digest: string;
  previous_release_id: string;
  expected_previous_release_digest: string;
}

export interface SetMiniAppEnabledRequest {
  miniapp_id: MiniAppId;
  expected_product_revision: number;
  expected_pointer_revision: number;
  expected_active_release_digest?: string;
  enabled: boolean;
}

export interface SetMiniAppPublishModeRequest {
  miniapp_id: MiniAppId;
  expected_product_revision: number;
  expected_pointer_revision: number;
  mode: MiniAppPublishMode;
}

export interface SetMiniAppServiceRunningRequest {
  miniapp_id: MiniAppId;
  expected_product_revision: number;
  expected_pointer_revision: number;
  expected_active_release_epoch: number;
  expected_active_release_digest: string;
  running: boolean;
}

export interface RetryMiniAppDeleteRequest {
  miniapp_id: MiniAppId;
  failed_operation_id: string;
  expected_operation_revision: number;
}

export interface TrashMiniAppRequest {
  miniapp_id: MiniAppId;
  expected_product_revision: number;
  expected_pointer_revision: number;
  expected_active_release_digest?: string;
}

export interface RestoreMiniAppRequest {
  miniapp_id: MiniAppId;
  expected_product_revision: number;
  expected_lifecycle: 'trashed';
  expected_pointer_revision: number;
}

export interface DeleteMiniAppRequest {
  miniapp_id: MiniAppId;
  expected_product_revision: number;
  expected_lifecycle: 'trashed';
  expected_pointer_revision: number;
  expected_active_release_digest?: string;
}

export interface RetryMiniAppServiceRequest {
  miniapp_id: MiniAppId;
  expected_product_revision: number;
  expected_pointer_revision: number;
  expected_active_release_epoch: number;
  expected_active_release_digest: string;
}

export interface OpenMiniAppSurfaceRequest {
  miniapp_id: MiniAppId;
}

export interface MiniAppSurfaceLaunchDescriptor {
  miniapp_id: MiniAppId;
  product_revision: number;
  release_id: string;
  expected_release_digest: string;
  active_release_epoch: number;
  surface_session_id: string;
  surface_generation: number;
  surface_capability: string;
  ui_entrypoint: string;
  kind: MiniAppKind;
}

export interface CloseMiniAppSurfaceRequest {
  miniapp_id: MiniAppId;
  surface_session_id: string;
  surface_capability: string;
}

export type MiniAppBridgeKvRequest =
  | { operation: 'get'; key: string }
  | { operation: 'set'; key: string; value: unknown }
  | { operation: 'delete'; key: string }
  | {
      operation: 'compare_and_swap';
      key: string;
      expected_revision?: number;
      value?: unknown;
    };

export interface MiniAppBridgeRequest {
  call_id: string;
  target:
    | {
        target: 'host_kv';
        request: MiniAppBridgeKvRequest;
      }
    | {
        target: 'service';
        method: string;
        payload: Record<string, unknown>;
      };
}

export type MiniAppKvResponse =
  | { outcome: 'value'; value?: unknown; revision?: number }
  | { outcome: 'written'; revision: number }
  | { outcome: 'deleted'; existed: boolean }
  | {
      outcome: 'compare_and_swap';
      applied: boolean;
      current_revision?: number;
    };

export interface MiniAppSurfaceBridgeRequest {
  miniapp_id: MiniAppId;
  surface_capability: string;
  active_release_epoch: number;
  expected_release_digest: string;
  request: MiniAppBridgeRequest;
}

File: crates/backend/nomifun-api-types/src/miniapp_platform.rs
//! Product-facing DTOs for the Phase M1 MiniApp Platform.
//!
//! Surface launch data contains only Host-consumable descriptors. MessageChannel
//! Bridge sessions and internal Service/storage handles remain private.

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::{
    CredentialSlotBindingDto, DurableOperationSummaryDto, PluginCapabilityContributionDto,
    PluginConfigSchemaDto, PluginConfigStateDto,
};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MiniAppKindDto {
    UiOnly,
    Service,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MiniAppServiceLifecycleDto {
    OnDemand,
    Continuous,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MiniAppLifecycleDto {
    Enabled,
    Disabled,
    Trashed,
    Deleting,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "state", rename_all = "snake_case", deny_unknown_fields)]
pub enum MiniAppServiceHealthDto {
    NotApplicable,
    Stopped,
    Starting {
        release_id: String,
        expected_release_digest: String,
    },
    Ready {
        release_id: String,
        expected_release_digest: String,
        started_at_ms: i64,
    },
    Failed {
        release_id: String,
        expected_release_digest: String,
        error_code: String,
    },
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct MiniAppReleaseRefDto {
    pub release_id: String,
    pub artifact_id: String,
    pub release_digest: String,
    pub manifest_digest: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct MiniAppReleasePointersDto {
    pub pointer_revision: u64,
    pub active_release_epoch: u64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ready: Option<MiniAppReleaseRefDto>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub active: Option<MiniAppReleaseRefDto>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub previous: Option<MiniAppReleaseRefDto>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct MiniAppSummaryDto {
    pub miniapp_id: String,
    pub product_revision: u64,
    pub display_name: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub icon_asset_id: Option<String>,
    pub kind: MiniAppKindDto,
    pub lifecycle: MiniAppLifecycleDto,
    pub releases: MiniAppReleasePointersDto,
    pub service_health: MiniAppServiceHealthDto,
    pub surface_available: bool,
    pub updated_at_ms: i64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct MiniAppLibraryResponseDto {
    pub library_revision: u64,
    pub miniapps: Vec<MiniAppSummaryDto>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MiniAppProjectSourceStateDto {
    Empty,
    Editable,
    RuntimeOnly,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct MiniAppServiceDescriptorDto {
    pub lifecycle: MiniAppServiceLifecycleDto,
    pub uses_files: bool,
    pub uses_private_database: bool,
    pub service_contract_digest: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MiniAppTestStatusDto {
    NotRequired,
    NotRun,
    Passed,
    Failed,
    NeedsTestInput,
    Stale,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct MiniAppReleaseTestDto {
    pub status: MiniAppTestStatusDto,
    pub release_id: String,
    pub expected_release_digest: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub receipt_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expected_service_run_key: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub issued_at_ms: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error_code: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct MiniAppReadyReleaseDto {
    pub release: MiniAppReleaseRefDto,
    pub project_build_generation: u64,
    pub created_at_ms: i64,
    pub kind: MiniAppKindDto,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub service: Option<MiniAppServiceDescriptorDto>,
    pub test: MiniAppReleaseTestDto,
    pub migration_count: u32,
    pub can_publish: bool,
    pub can_auto_publish: bool,
    #[serde(default)]
    pub blocking_reasons: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct MiniAppWorkshopDto {
    pub miniapp: MiniAppSummaryDto,
    pub project_id: String,
    pub project_revision: u64,
    pub source_state: MiniAppProjectSourceStateDto,
    pub build_generation: u64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_snapshot_digest: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub dependency_lock_digest: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ready: Option<MiniAppReadyReleaseDto>,
    pub config_schema: PluginConfigSchemaDto,
    pub config: PluginConfigStateDto,
    pub credential_bindings_revision: u64,
    pub credential_slots: Vec<CredentialSlotBindingDto>,
    pub capabilities: Vec<PluginCapabilityContributionDto>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub active_operation: Option<DurableOperationSummaryDto>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct MiniAppSurfaceLaunchDescriptorDto {
    pub miniapp_id: String,
    pub product_revision: u64,
    pub release_id: String,
    pub expected_release_digest: String,
    pub active_release_epoch: u64,
    pub ui_asset_id: String,
    pub ui_entrypoint: String,
    pub kind: MiniAppKindDto,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct CreateMiniAppProjectRequest {
    pub expected_library_revision: u64,
    pub display_name: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    pub kind: MiniAppKindDto,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct BuildMiniAppRequest {
    pub miniapp_id: String,
    pub expected_product_revision: u64,
    pub project_id: String,
    pub expected_project_revision: u64,
    pub expected_build_generation: u64,
    pub expected_source_snapshot_digest: String,
    pub expected_dependency_lock_digest: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct CancelMiniAppBuildRequest {
    pub expected_operation_revision: u64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct TestMiniAppReleaseRequest {
    pub miniapp_id: String,
    pub expected_product_revision: u64,
    pub expected_pointer_revision: u64,
    pub project_id: String,
    pub expected_project_revision: u64,
    pub expected_build_generation: u64,
    pub release_id: String,
    pub expected_release_digest: String,
    pub expected_config_revision: u64,
    pub expected_credential_bindings_revision: u64,
    pub resolved_test_input_digest: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct PublishMiniAppRequest {
    pub miniapp_id: String,
    pub expected_product_revision: u64,
    pub expected_pointer_revision: u64,
    pub expected_active_release_epoch: u64,
    pub ready_release_id: String,
    pub expected_ready_release_digest: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expected_active_release_digest: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expected_service_test_receipt_id: Option<String>,
    pub acknowledge_test_warning: bool,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MiniAppPublishModeDto {
    Manual,
    AutoUiOnly,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct SetMiniAppPublishModeRequest {
    pub miniapp_id: String,
    pub expected_product_revision: u64,
    pub expected_pointer_revision: u64,
    pub mode: MiniAppPublishModeDto,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct DiscardMiniAppReadyReleaseRequest {
    pub miniapp_id: String,
    pub expected_product_revision: u64,
    pub expected_pointer_revision: u64,
    pub ready_release_id: String,
    pub expected_ready_release_digest: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct RollbackMiniAppRequest {
    pub miniapp_id: String,
    pub expected_product_revision: u64,
    pub expected_pointer_revision: u64,
    pub expected_active_release_epoch: u64,
    pub expected_current_release_digest: String,
    pub previous_release_id: String,
    pub expected_previous_release_digest: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct SetMiniAppEnabledRequest {
    pub miniapp_id: String,
    pub expected_product_revision: u64,
    pub expected_pointer_revision: u64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expected_active_release_digest: Option<String>,
    pub enabled: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct TrashMiniAppRequest {
    pub miniapp_id: String,
    pub expected_product_revision: u64,
    pub expected_pointer_revision: u64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expected_active_release_digest: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct RestoreMiniAppRequest {
    pub miniapp_id: String,
    pub expected_product_revision: u64,
    pub expected_lifecycle: MiniAppLifecycleDto,
    pub expected_pointer_revision: u64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct RetryMiniAppServiceRequest {
    pub miniapp_id: String,
    pub expected_product_revision: u64,
    pub expected_pointer_revision: u64,
    pub expected_active_release_epoch: u64,
    pub expected_active_release_digest: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct SetMiniAppServiceRunningRequest {
    pub miniapp_id: String,
    pub expected_product_revision: u64,
    pub expected_pointer_revision: u64,
    pub expected_active_release_epoch: u64,
    pub expected_active_release_digest: String,
    pub running: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct DeleteMiniAppRequest {
    pub miniapp_id: String,
    pub expected_product_revision: u64,
    pub expected_lifecycle: MiniAppLifecycleDto,
    pub expected_pointer_revision: u64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expected_active_release_digest: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MiniAppShareContentDto {
    ReadyRelease,
    ActiveRelease,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ShareMiniAppRequest {
    pub miniapp_id: String,
    pub expected_product_revision: u64,
    pub expected_pointer_revision: u64,
    pub content: MiniAppShareContentDto,
    pub release_id: String,
    pub expected_release_digest: String,
    pub destination_path: String,
    pub include_source: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ImportMiniAppShareRequest {
    pub expected_library_revision: u64,
    pub source_path: String,
    pub expected_bundle_digest: String,
    pub expected_release_digest: String,
    pub display_name: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ImportMiniAppArtifactRequest {
    pub expected_library_revision: u64,
    pub source_path: String,
    pub expected_artifact_digest: String,
    pub display_name: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ExportMiniAppBackupRequest {
    pub miniapp_id: String,
    pub expected_product_revision: u64,
    pub expected_lifecycle: MiniAppLifecycleDto,
    pub expected_pointer_revision: u64,
    pub expected_config_revision: u64,
    pub expected_credential_bindings_revision: u64,
    pub destination_path: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ImportMiniAppBackupRequest {
    pub expected_library_revision: u64,
    pub source_path: String,
    pub expected_backup_metadata_digest: String,
    pub display_name: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ConfigureMiniAppRequest {
    pub miniapp_id: String,
    pub expected_product_revision: u64,
    pub expected_pointer_revision: u64,
    pub expected_config_revision: u64,
    pub expected_schema_digest: String,
    pub values: Value,
    pub credential_bindings: std::collections::BTreeMap<String, Option<String>>,
    pub expected_credential_bindings_revision: u64,
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn serialized_field_names(value: &Value, output: &mut Vec<String>) {
        match value {
            Value::Object(fields) => {
                for (key, value) in fields {
                    output.push(key.clone());
                    serialized_field_names(value, output);
                }
            }
            Value::Array(values) => {
                for value in values {
                    serialized_field_names(value, output);
                }
            }
            _ => {}
        }
    }

    #[test]
    fn surface_launch_has_no_bridge_or_localhost_wire() {
        let value = serde_json::to_value(MiniAppSurfaceLaunchDescriptorDto {
            miniapp_id: "miniapp-1".into(),
            product_revision: 3,
            release_id: "release-2".into(),
            expected_release_digest: "a".repeat(64),
            active_release_epoch: 8,
            ui_asset_id: "artifact-2".into(),
            ui_entrypoint: "ui/index.html".into(),
            kind: MiniAppKindDto::Service,
        })
        .unwrap();

        let mut fields = Vec::new();
        serialized_field_names(&value, &mut fields);
        for forbidden in [
            "bridge",
            "message_channel",
            "localhost",
            "port",
            "host_generation",
            "service_run_key",
        ] {
            assert!(
                fields.iter().all(|field| !field.contains(forbidden)),
                "internal field leaked: {forbidden}"
            );
        }
        assert_eq!(value["kind"], "service");
    }

    #[test]
    fn miniapp_requests_reject_unknown_and_legacy_fields() {
        let error = serde_json::from_value::<BuildMiniAppRequest>(json!({
            "miniapp_id": "miniapp-1",
            "expected_product_revision": 3,
            "project_id": "project-1",
            "expected_project_revision": 4,
            "expected_build_generation": 5,
            "expected_source_snapshot_digest": "a".repeat(64),
            "expected_dependency_lock_digest": "b".repeat(64),
            "html": "<html></html>"
        }))
        .unwrap_err();

        assert!(error.to_string().contains("unknown field"));
    }

    #[test]
    fn publish_and_rollback_bind_exact_release_pointers() {
        let publish = serde_json::to_value(PublishMiniAppRequest {
            miniapp_id: "miniapp-1".into(),
            expected_product_revision: 3,
            expected_pointer_revision: 5,
            expected_active_release_epoch: 7,
            ready_release_id: "release-2".into(),
            expected_ready_release_digest: "a".repeat(64),
            expected_active_release_digest: Some("b".repeat(64)),
            expected_service_test_receipt_id: Some("receipt-1".into()),
            acknowledge_test_warning: false,
        })
        .unwrap();
        let rollback = serde_json::to_value(RollbackMiniAppRequest {
            miniapp_id: "miniapp-1".into(),
            expected_product_revision: 4,
            expected_pointer_revision: 6,
            expected_active_release_epoch: 8,
            expected_current_release_digest: "a".repeat(64),
            previous_release_id: "release-1".into(),
            expected_previous_release_digest: "b".repeat(64),
        })
        .unwrap();

        assert_eq!(publish["expected_pointer_revision"], 5);
        assert_eq!(publish["expected_active_release_epoch"], 7);
        assert_eq!(rollback["expected_pointer_revision"], 6);
        assert_eq!(rollback["expected_active_release_epoch"], 8);
        assert!(publish.get("conversation_id").is_none());
        assert!(publish.get("guid_mode").is_none());
    }

    #[test]
    fn miniapp_credential_binding_wire_contains_references_only() {
        let request = ConfigureMiniAppRequest {
            miniapp_id: "miniapp-1".into(),
            expected_product_revision: 3,
            expected_pointer_revision: 5,
            expected_config_revision: 2,
            expected_schema_digest: "a".repeat(64),
            values: json!({"region": "us-east"}),
            credential_bindings: std::collections::BTreeMap::from([(
                "provider".into(),
                Some("credential-1".into()),
            )]),
            expected_credential_bindings_revision: 4,
        };

        let value = serde_json::to_value(request).unwrap();
        let mut fields = Vec::new();
        serialized_field_names(&value, &mut fields);
        assert!(fields.iter().all(|field| !field.contains("secret")));
        assert!(fields.iter().all(|field| !field.contains("plaintext")));
        assert_eq!(value["credential_bindings"]["provider"], "credential-1");
    }
}
